src/VisualStudio/Core/Def/TaskList/ExternalErrorDiagnosticUpdateSource.cs

```diff
@@ -5,132 +5,104 @@
+    /// </summary>
+    private IDiagnosticManagerService? _diagnosticManagerService;
 
     // Gate for concurrent access and fields guarded with this gate.
     private readonly object _gate = new();
     private InProgressState? _stateDoNotAccessDirectly;
-    private readonly CancellationSeries _activeCancellationSeriesDoNotAccessDirectly = new();
-
-    /// <summary>
-    /// Latest diagnostics reported during current or last build.
-    /// These are not the de-duped build/live diagnostics, but the actual diagnostics from build.
-    /// They are directly used by the "Build only" error list setting.
-    /// </summary>
-    private ImmutableArray<DiagnosticData> _lastBuiltResult = ImmutableArray<DiagnosticData>.Empty;
 
+    [ImportingConstructor]
+    [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
     public ExternalErrorDiagnosticUpdateSource(
         VisualStudioWorkspace workspace,
         IDiagnosticAnalyzerService diagnosticService,
-        IGlobalOperationNotificationService notificationService,
         IAsynchronousOperationListenerProvider listenerProvider,
+        [Import(typeof(SVsFullAccessServiceBroker))] IServiceBroker serviceBroker,
         IThreadingContext threadingContext)
-        : this(workspace, diagnosticService, notificationService, listenerProvider.GetListener(FeatureAttribute.ErrorList), threadingContext.DisposalToken)
-    {
-    }
-
-    /// <summary>
-    /// internal for testing
-    /// </summary>
-    internal ExternalErrorDiagnosticUpdateSource(
-        Workspace workspace,
-        IDiagnosticAnalyzerService diagnosticService,
-        IGlobalOperationNotificationService notificationService,
-        IAsynchronousOperationListener listener,
-        CancellationToken disposalToken)
     {
-        // use queue to serialize work. no lock needed
-        _taskQueue = new TaskQueue(listener, TaskScheduler.Default);
-        _disposalToken = disposalToken;
-
+        _disposalToken = threadingContext.DisposalToken;
         _workspace = workspace;
-        _workspace.WorkspaceChanged += OnWorkspaceChanged;
-
         _diagnosticService = diagnosticService;
-        _buildOnlyDiagnosticsService = _workspace.Services.GetRequiredService<IBuildOnlyDiagnosticsService>();
+        _listener = listenerProvider.GetListener(FeatureAttribute.ErrorList);
+
+        _serviceBroker = serviceBroker;
+        _taskQueue = new AsyncBatchingWorkQueue<Func<CancellationToken, Task>>(
+            TimeSpan.Zero,
+            processBatchAsync: ProcessTaskQueueItemsAsync,
+            _listener,
+            _disposalToken
+        );
+    }
 
-        _notificationService = notificationService;
+    private async ValueTask ProcessTaskQueueItemsAsync(ImmutableSegmentedList<Func<CancellationToken, Task>> list, CancellationToken cancellationToken)
+    {
+        foreach (var workItem in list)
+            await workItem(cancellationToken).ConfigureAwait(false);
```

Comment:
nit: we *might* want to wrap these with try/catch/reportnfw, so that we do the remainder of work if one dies for non-cancellation reasons.  but i don't feel strongly.